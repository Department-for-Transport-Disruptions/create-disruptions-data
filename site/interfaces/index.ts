import { NextApiRequest } from "next";
import { z } from "zod";
import { ServerResponse } from "http";
import { Service, Stop } from "../schemas/consequence.schema";
import { ConsequenceType } from "../schemas/type-of-consequence.schema";

export interface ErrorInfo {
    errorMessage: string;
    id: string;
    userInput?: string;
}

export interface ResponseWithLocals extends ServerResponse {
    locals: {
        nonce: string;
        csrfToken: string;
    };
}

export interface CookiePolicy {
    essential: boolean;
    usage: boolean;
}

export interface CookiesApiRequest extends NextApiRequest {
    body: {
        tracking: "on" | "off";
    };
}

export interface FormBase<T> {
    value?: string;
    display: string;
    displaySize?: "s" | "m" | "l" | "xl";
    inputName: Extract<keyof T, string>;
    initialErrors?: ErrorInfo[];
    stateUpdater: (change: string, field: keyof T) => void;
    schema?: z.ZodTypeAny;
}

export interface DisplayValuePair<T = string> {
    display: string;
    value: T;
    checked?: boolean;
}

export interface Feedback {
    question: string;
    answer: string;
}

export interface TestInputs {
    field1: string;
    field2: boolean;
    field3: string;
    field4: string;
}

export interface PageState<T> {
    errors: ErrorInfo[];
    inputs: T;
}

export interface SocialMediaPost {
    messageToAppear: string;
    publishDate: string;
    publishTime: string;
    accountToPublish: string;
}

export interface CreateConsequenceProps<T> {
    initialPageState: PageState<Partial<T>>;
    previousConsequenceInformation: ConsequenceType;
    initialServices?: Service[];
    initialStops?: Stop[];
    allStops?: Stop[]
}
