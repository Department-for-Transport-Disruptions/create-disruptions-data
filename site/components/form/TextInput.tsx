import kebabCase from "lodash/kebabCase";
import { ReactElement, useEffect, useState } from "react";
import FormElementWrapper, { FormGroupWrapper } from "./FormElementWrapper";
import { ErrorInfo, FormBase } from "../../interfaces";

interface TextInputProps<T> extends FormBase<T> {
    widthClass: string;
    maxLength: number;
    textArea?: boolean;
    rows?: number;
    hint?: string;
    isPassword?: boolean;
    resetError?: boolean;
}

const TextInput = <T extends object>({
    value,
    display,
    inputName,
    displaySize = "s",
    initialErrors = [],
    widthClass,
    maxLength,
    textArea = false,
    rows,
    hint,
    stateUpdater,
    isPassword,
    resetError = false,
}: TextInputProps<T>): ReactElement => {
    const [errors, setErrors] = useState<ErrorInfo[]>(initialErrors);
    const inputId = kebabCase(inputName);

    useEffect(() => {
        if (resetError) {
            setErrors([]);
        }
    }, [resetError, setErrors]);

    return (
        <FormGroupWrapper errorIds={[inputName]} errors={errors}>
            <div className="govuk-form-group" id={inputId}>
                <label className={`govuk-label govuk-label--${displaySize}`} htmlFor={`${inputId}-input`}>
                    {display}
                </label>
                {hint ? (
                    <div id={`${inputId}-hint`} className="govuk-hint">
                        {hint}
                    </div>
                ) : null}
                <FormElementWrapper errors={errors} errorId={inputName} errorClass="govuk-input--error">
                    {textArea ? (
                        <textarea
                            className={`govuk-textarea ${widthClass}`}
                            name={inputName}
                            id={`${inputId}-input`}
                            rows={rows}
                            maxLength={maxLength}
                            value={value}
                            onChange={(e) => {
                                stateUpdater(e.target.value, inputName);
                            }}
                            aria-describedby={!!hint ? `${inputId}-hint` : undefined}
                        />
                    ) : (
                        <input
                            className={`govuk-input ${widthClass}`}
                            id={`${inputId}-input`}
                            name={inputName}
                            type={isPassword ? "password" : "text"}
                            maxLength={maxLength}
                            value={value}
                            onChange={(e) => {
                                stateUpdater(e.target.value, inputName);
                            }}
                            aria-describedby={!!hint ? `${inputId}-hint` : undefined}
                        />
                    )}
                </FormElementWrapper>
            </div>
        </FormGroupWrapper>
    );
};

export default TextInput;
