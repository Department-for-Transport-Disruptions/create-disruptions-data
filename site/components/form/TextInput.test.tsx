import renderer from "react-test-renderer";
import { describe, it, expect, vi } from "vitest";
import TextInput from "./TextInput";
import { TestInputs } from "../../interfaces";

describe("TextInput", () => {
    it("should render correctly with no errors", () => {
        const tree = renderer
            .create(
                <TextInput<TestInputs>
                    inputName="field1"
                    display="Test Field"
                    stateUpdater={vi.fn()}
                    widthClass="w-3/4"
                    maxLength={50}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with errors", () => {
        const tree = renderer
            .create(
                <TextInput<TestInputs>
                    initialErrors={[{ errorMessage: "There was an error", id: "summary" }]}
                    inputName="field1"
                    display="Test Field"
                    stateUpdater={vi.fn()}
                    widthClass="w-3/4"
                    maxLength={50}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly as a text area with rows", () => {
        const tree = renderer
            .create(
                <TextInput<TestInputs>
                    inputName="field1"
                    display="Test Field"
                    stateUpdater={vi.fn()}
                    widthClass="w-3/4"
                    maxLength={50}
                    textArea
                    rows={3}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
