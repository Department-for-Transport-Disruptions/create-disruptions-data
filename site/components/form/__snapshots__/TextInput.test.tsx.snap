// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextInput > should render correctly as a text area with rows 1`] = `
<div
  className="govuk-form-group"
>
  <div
    className="govuk-form-group"
    id="field-1"
  >
    <label
      className="govuk-label govuk-label--s"
      htmlFor="field-1-input"
    >
      Test Field
    </label>
    <div
      className=""
    >
      <textarea
        className="govuk-textarea w-3/4"
        id="field-1-input"
        maxLength={50}
        name="field1"
        onChange={[Function]}
        rows={3}
      />
    </div>
  </div>
</div>
`;

exports[`TextInput > should render correctly with errors 1`] = `
<div
  className="govuk-form-group"
>
  <div
    className="govuk-form-group"
    id="field-1"
  >
    <label
      className="govuk-label govuk-label--s"
      htmlFor="field-1-input"
    >
      Test Field
    </label>
    <div
      className=""
    >
      <input
        className="govuk-input w-3/4"
        id="field-1-input"
        maxLength={50}
        name="field1"
        onChange={[Function]}
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`TextInput > should render correctly with no errors 1`] = `
<div
  className="govuk-form-group"
>
  <div
    className="govuk-form-group"
    id="field-1"
  >
    <label
      className="govuk-label govuk-label--s"
      htmlFor="field-1-input"
    >
      Test Field
    </label>
    <div
      className=""
    >
      <input
        className="govuk-input w-3/4"
        id="field-1-input"
        maxLength={50}
        name="field1"
        onChange={[Function]}
        type="text"
      />
    </div>
  </div>
</div>
`;
